libcontainer/nsenter/nsexec.c

```diff
@@ -1,471 +1,753 @@
+
+		fd = open(path, O_RDONLY);
+		if (fd < 0)
+			bail("failed to open %s", namespace);
+
+		ns->fd = fd;
+		ns->ns = nsflag(namespace);
+		strncpy(ns->path, path, PATH_MAX);
+	} while ((namespace = strtok_r(NULL, ",", &saveptr)) != NULL);
 
-	// Send the child pid back to our parent
-	len = snprintf(buf, sizeof(buf), "{ \"pid\" : %d }\n", childpid);
-	if ((len < 0) || (write(pipenum, buf, len) != len)) {
-		pr_perror("Unable to send a child pid");
-		kill(childpid, SIGKILL);
-		exit(1);
-	}
+	/*
+	 * The ordering in which we join namespaces is important. We should
+	 * always join the user namespace *first*. This is all guaranteed
+	 * from the container_linux.go side of this, so we're just going to
+	 * follow the order given to us.
+	 */
 
-	exit(0);
-}
+	for (i = 0; i < num; i++) {
+		struct namespace_t ns = namespaces[i];
 
-static struct nsenter_config process_nl_attributes(int pipenum, char *data, int data_size)
-{
-	struct nsenter_config	config	    = {0};
-	struct nlattr		*nlattr;
-	int			payload_len;
-	int			start       = 0;
-
-	config.consolefd = -1;
-	while (start < data_size) {
-		nlattr = (struct nlattr *)(data + start);
-		start += NLA_HDRLEN;
-		payload_len = nlattr->nla_len - NLA_HDRLEN;
-
-		if (nlattr->nla_type == CLONE_FLAGS_ATTR) {
-			config.cloneflags = readint32(data + start);
-		} else if (nlattr->nla_type == CONSOLE_PATH_ATTR) {
-			// get the console path before setns because it may
-			// change mnt namespace
-			config.consolefd = open(data + start, O_RDWR);
-			if (config.consolefd < 0) {
-				pr_perror("Failed to open console %s",
-					  data + start);
-				exit(1);
-			}
-		} else if (nlattr->nla_type == NS_PATHS_ATTR) {
-			// if custom namespaces are required, open all
-			// descriptors and perform setns on them
-			int	i, j;
-			int	nslen = num_namespaces(data + start);
-			int	fds[nslen];
-			char	*nslist[nslen];
-			char	*ns;
-			char	*saveptr = NULL;
-
-			for (i = 0; i < nslen; i++) {
-				char *str = NULL;
-
-				if (i == 0) {
-					str = data + start;
-				}
-				ns = strtok_r(str, ",", &saveptr);
-				if (ns == NULL) {
-					break;
-				}
-				fds[i] = open(ns, O_RDONLY);
-				if (fds[i] == -1) {
-					for (j = 0; j < i; j++) {
-						close(fds[j]);
-					}
-					pr_perror("Failed to open %s", ns);
-					exit(1);
-				}
-				nslist[i] = ns;
-			}
-
-			for (i = 0; i < nslen; i++) {
-				if (setns(fds[i], 0) != 0) {
-					pr_perror("Failed to setns to %s", nslist[i]);
-					exit(1);
-				}
-				close(fds[i]);
-			}
-		} else if (nlattr->nla_type == UIDMAP_ATTR) {
-			config.uidmap     = data + start;
-			config.uidmap_len = payload_len;
-		} else if (nlattr->nla_type == GIDMAP_ATTR) {
-			config.gidmap     = data + start;
-			config.gidmap_len = payload_len;
-		} else if (nlattr->nla_type == SETGROUP_ATTR) {
-			config.is_setgroup = readint8(data + start);
-		} else {
-			pr_perror("Unknown netlink message type %d",
-				  nlattr->nla_type);
-			exit(1);
-		}
+		if (setns(ns.fd, ns.ns) < 0)
```

Comment:
You need to set uid, gid, groups from a target user namespace before switching into it, otherwise you will have a security issue.
setresgid(...);
setresuid(...);
setns(CLONE_NEWUSER);

https://github.com/xemul/criu/commit/6e1726f8fb7c5a22538c61ef3828c375cc44d0d1 is an example how the same problem was fixed in CRIU.